ENH Add N-D fitting and evaluation methods to the chebyshev.py and polynomial.py library

pbrod opened the pull request:
Currently there are no functions in chebyshev.py/ polynomial.py that do 
- N-dimensional fitting for N>1
- N-dimensional evaluation for N>3
- Construct Vandermonde matrix for N>3

This submission is a first implementation to fill the above mentioned gap.
Functions added to chebyshev.py and polynomial.py are:
chebvalnd, chebgridnd, chebfitnd, chebfitfunc, chebvandernd
polyvalnd, polygridnd, polyfitnd, polyfitfunc, polyvandernd
and supporting tests
# Examples on use

```
>>> import numpy as np
>>> import numpy.polynomial.chebyshev as cheb
>>> import matplotlib.pyplot as plt
>>> n = 51
>>> xorder, yorder = n-1, n-1
>>> x = cheb.chebpts1(n)
>>> xgrid, ygrid = np.meshgrid(x, x)
>>> def f(x, y): return np.exp(-x**2-6*y**2)
>>> zgrid = f(xgrid, ygrid)
>>> dcoeff2 = cheb.chebfitnd((xgrid, ygrid), zgrid, [xorder,yorder])
>>> dcoeff = cheb.chebfitfunc(f, n=(xorder+1,yorder+1))
>>> np.allclose(dcoeff2, dcoeff)
True
```

Visualize the significant coefficients

```
>>> plt.figure()
>>> plt.spy(np.abs(dcoeff)>1e-13)
>>> plt.title('Coefficients > 10^-13')
```

![figure_1](https://cloud.githubusercontent.com/assets/1193469/8652374/b7e5ce6e-297e-11e5-8bee-429075430eb2.png) 

Evaluate on finer grid

```
>>> xi = np.linspace(-1, 1, 151)
>>> Xi,Yi = np.meshgrid(xi, xi)
>>> Zi = f(Xi, Yi)
```

Compute residuals

```
>>> zzi = cheb.chebvalnd(dcoeff, Xi, Yi)
>>> devi = Zi - zzi

>>> zz = cheb.chebvalnd(dcoeff, xgrid, ygrid)
>>> dev = zgrid - zz
```

and plot them

```
>>> plt.figure() 
>>> plt.contourf(xgrid, ygrid, np.abs(dev))
>>> plt.colorbar()
>>> plt.title('Error at the chebfitnd-points')

>>> plt.figure()
>>> plt.contourf(Xi, Yi, np.abs(devi))
>>> plt.colorbar()
>>> plt.title('Error on a fine grid')
>>> plt.show('hold')
```

![figure_2](https://cloud.githubusercontent.com/assets/1193469/8652376/b7e91448-297e-11e5-9a4b-73925c7dc543.png)
![figure_3](https://cloud.githubusercontent.com/assets/1193469/8652375/b7e60ef6-297e-11e5-954a-bca4be26cf2e.png)

charris commented:
You should squash the commits and follow the guidelines in `doc/source/dev/gitwash/development_workflow.rst` for the messages. Also, `doc/source/reference/routines.polynomials.chebyshev.rst` will need an update before this is merged, and perhaps the `doc/release/1.10.0-notes.rst` also if this is merged before the 1.10.0 branch.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1411,6 +1412,111 @@ def chebgrid3d(x, y, z, c):
     return c
 
```
Need to add the new functions to `__all__`.

charris replied:
And also in the module documentation.

pbrod replied:
Ok. This is done. I have updated the files to according to your comments.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1411,6 +1412,111 @@ def chebgrid3d(x, y, z, c):
+    try:
+        xi = np.array(xi, copy=0)
+    except:
+        raise ValueError('x, y, z are incompatible')
```
The error message should refer to the general `(x1, x2,...,xn)` shape for the point coordinates.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1594,6 +1700,70 @@ def chebvander3d(x, y, z, deg):
+    points `(x1, x2, ..., xn)`. If `l, m, n` are the given degrees in
+    `x1, x2, x3`, then The pseudo-Vandermonde matrix is defined by
+
+    .. math:: V[..., (m+1)(n+1)i + (n+1)j + k] = T_i(x1)*T_j(x2)*T_k(x3),
```
This is not correct for the general nd case. Might be tricky to express it for the general case.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1594,6 +1700,70 @@ def chebvander3d(x, y, z, deg):
+    shape0 = xi[0].shape
+    s0 = (1,) * ndim
+    vxi = [chebvander(x, d).reshape(shape0 + s0[:i] + (-1,) + s0[i + 1::])
+            for i, (d, x) in enumerate(zip(ideg, xi))]
```
The `f` should align with the upper `c`. Might want to run pep8 and pyflakes on this PR.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1594,6 +1700,70 @@ def chebvander3d(x, y, z, deg):
+    xi = np.array(xi, copy=0) + 0.0
+    shape0 = xi[0].shape
+    s0 = (1,) * ndim
+    vxi = [chebvander(x, d).reshape(shape0 + s0[:i] + (-1,) + s0[i + 1::])
```
This might read better as a for loop with several lines rather than a list comprehension.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1594,6 +1700,70 @@ def chebvander3d(x, y, z, deg):
+    vxi = [chebvander(x, d).reshape(shape0 + s0[:i] + (-1,) + s0[i + 1::])
+            for i, (d, x) in enumerate(zip(ideg, xi))]
+
+    v = reduce(np.multiply, vxi)
```
And this could be part of the for loop (stepwise) and avoid the use of `reduce`.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1766,6 +1936,255 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+    coefficients to be solved for, `w` are the weights, and `y` are the
+    observed values.  This equation is then solved using the singular value
+    decomposition of `V`.
+    If some of the singular values of `V` are so small that they are
```
Blank lines before paragraphs. If paragraph not intended, should make the line lengths more consistent, `gq` over selection in vim.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1766,6 +1936,255 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+           http://en.wikipedia.org/wiki/Curve_fitting
+    Examples
+    --------
+    """
```
Blank line before `"""`. Not sure if you should have `Examples` if there are none.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -90,9 +90,10 @@
+from functools import reduce
 from . import polyutils as pu
 from ._polybase import ABCPolyBase
+# from scipy.fftpack import dct, idct as _idct
```
Don't want a scipy dependency.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1766,6 +1936,255 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+        return c
+
+
+# def chebfit_dct(f, n=(10, ), domain=None):
```
The name is too method specific, i.e., `dct`. Maybe `chebfitfunc` or some such. You can use the Gauss points here with the weights, a method which is probably not much slower than the dct for arbitrary values of `n` that don't have nice factorizations and will generalize to polynomials other than Chebyshev.

charris commented:
I like this in general. The tests are a bit skimpy and some of my style comments apply more generally. Once this is settled, it should be extended to the remaining polynomials, although power series will be a bit of an exception.

I've thought a bit about making classes for some of this similar to the Polynomial etc. classes but with fewer features. Two other generalization that might be useful in practice are vector valued functions and gradients and Jacobians. If you have any thoughts along those lines I'd like to hear them.

charris commented:
I'm curious if you actually have a use case for the > 3d case?

pbrod committed:
ENH: Added N-D fitting and evaluation methods:
Added:
chebvalnd, chebgridnd, chebfitnd, chebfitfunc, chebvandernd
polvalnd, polygridnd, polyfitnd, polyfitfunc, polyvandernd

Also added supporting tests

charris commented:
The second commit should be a bit more specific and begin with `DOC:`, or you could just squash it into the first...

Now begins the drudgery. We try to keep all 6 polynomials in sync, so functions added to one also need to be added to the remaining 5 and tested. Yep, it's a PITA ;)

charris commented:
If this goes in, it will also need a mention in `doc/release/1.11.0-notes.rst`.

charris commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1411,6 +1418,113 @@ def chebgrid3d(x, y, z, c):
+    If `c` has fewer than N dimensions, ones are implicitly appended to its
+    shape to make it N-D. The shape of the result will be c.shape[3:] +
+    x1.shape.
+
```
Needs `.. versionadded:: 1.11.0` here, followed by a blank line.

charris replied:
Grep for other examples.

pbrod commented:
Added doc/release/1.11.0-notes.rst and .. versionadded:: 1.11.0.

seberg commented:
@pbrod, I think you have to update your master and rebase, it seems to me you created a new 1.11 release notes file. This already exists.

pbrod commented:
Ok. Thanks.

homu commented:
:umbrella: The latest upstream changes (presumably #6688) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #7735) made this pull request unmergeable. Please resolve the merge conflicts.

woodscn commented:
For what it's worth, I was just thinking about trying to implement something like this. I'm also interested in evaluation for N>3, i.e. x,y,z,t; or state spaces, etc. 

What is the current status of this PR?

charris commented:
IIRC, it was in decent shape. I was thinking of taking a further look at it in the next few days.

woodscn commented:
Looking at this more closely, I'm not sure if it does what I'm looking for. Essentially, I'm after something like a MultiVariatePolynomial helper class, with things like derivatives, integrals, +, *, etc.

charris commented:
@pbrod This is old but I still think this is useful. Want to pursue this?

charris commented:
@woodscn I agree that that would be useful. If you are motivated in that direction you may want to put together a PR.

Qiyu8 commented:
@pbrod Are you still working on this? looks valuable if you can solve the confict 😄

rossbar commented:
There is another attempt at expanding polynomial fitting to ND in #14151

scratchmex commented:
ping @pbrod. I think this is very valuable. Do you have any opinion or comments about the difference between this and #14151 as mentioned by rossbar?

pbrod commented:
I am closing this as  #14151 does the same and more.
